### PR TITLE
Allocation via virtual functions

### PR DESCRIPTION
**Pull request by orazve**

### CODE REVIEW

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/debug.cpp [hunk 1]
```diff
@@ -18,8 +18,8 @@ void pa_bit(char *str, std::uint8_t **a, size_t n) {
         std::cout << "    NULL";
         return;
     }
-    for (int i = 0; i < n; i++) {
-        for (int j = 0; j < n; j++) {
+    for (size_t i = 0; i < n; i++) {
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/debug.cpp:21]: I assume that this code to be removed in the future

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp [hunk 2]
```diff
@@ -103,8 +105,7 @@ graph::graph(const dal::preview::detail::topology<std::int32_t>& t,
         else {
             graph_data_storage.plist_data->degree[i] = degree;
             if (degree > 0) {
-                graph_data_storage.plist_data->data[i] =
-                    static_cast<std::int64_t*>(_mm_malloc(sizeof(std::int64_t) * degree, 64));
+                graph_data_storage.plist_data->data[i] = _allocator.allocate<std::int64_t>(degree);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp]: You changed memory alignment here and in other lines.

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp [hunk 3]
```diff
@@ -196,10 +195,13 @@ graph_data::~graph_data() {
 }
 
 graph::~graph() {
-    if (!external_data) {
+    if (external_data) {
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp:198]: Looks like it was correct logic that you changed. Are you sure?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp [hunk 4]
```diff
@@ -233,15 +235,28 @@ void graph::delete_bit_arrays() {
     if (p_edges_bit != nullptr) {
         for (std::int64_t i = 0; i < n; i++) {
             if (p_edges_bit[i] != nullptr) {
-                _mm_free(p_edges_bit[i]);
+                _allocator.deallocate<std::uint8_t>(p_edges_bit[i], 0);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp]: Why 0? It should be aligned with argument of allocate

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp [hunk 5]
```diff
@@ -233,15 +235,28 @@ void graph::delete_bit_arrays() {
...
+        _allocator.deallocate<std::uint8_t*>(p_edges_bit, n);
         p_edges_bit = nullptr;
     }
 }
 
+void graph::delete_list_arrays() {
+    if (p_edges_list != nullptr) {
+        for (std::int64_t i = 0; i < n; i++) {
+            if (p_edges_list[i] != nullptr) {
+                _allocator.deallocate<std::int64_t>(p_edges_list[i], 0);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp]: Again 0

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp [hunk 6]
```diff
@@ -430,30 +445,36 @@ graph_status bit_vector::unset_bit(std::uint8_t* result_vector, const std::int64
...
     n = vector_size;
-    vector = static_cast<std::uint8_t*>(_mm_malloc(sizeof(std::uint8_t) * n, 64));
+    vector = _allocator.allocate<std::uint8_t>(n);
     set(n, vector);
 }
 
-bit_vector::bit_vector(const std::int64_t vector_size, const std::uint8_t byte_val) {
+bit_vector::bit_vector(const std::int64_t vector_size,
+                       const std::uint8_t byte_val,
+                       inner_alloc _allocator)
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp]: Reference?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp [hunk 7]
```diff
@@ -812,29 +837,30 @@ graph_input_list_data::graph_input_list_data(graph_input_data* input_data) {
 graph_input_list_data::~graph_input_list_data() {
     for (int64_t i = 0; i < vertex_count; i++) {
         if (data[i] != nullptr) {
-            _mm_free(data[i]);
+            _allocator.deallocate<std::int64_t>(data[i], 0);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp]: Again 0

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp [hunk 8]
```diff
@@ -847,20 +873,20 @@ graph_input_bit_data::graph_input_bit_data(graph_input_data* input_data) {
...
+        data[i] = _allocator.allocate<std::uint8_t>(bit_array_size);
         bit_vector::set(bit_array_size, data[i]);
     }
 }
 
 graph_input_bit_data::~graph_input_bit_data() {
     for (int64_t i = 0; i < vertex_count; i++) {
         if (data[i] != nullptr) {
-            _mm_free(data[i]);
+            _allocator.deallocate<std::uint8_t>(data[i], 0);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.cpp]: Again 0

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp [hunk 9]
```diff
@@ -12,6 +13,48 @@
 
 namespace oneapi::dal::preview::subgraph_isomorphism::detail {
 
+struct byte_alloc_iface {
+    using byte_t = char;
+    virtual byte_t* allocate(std::int64_t n) = 0;
+    virtual void deallocate(byte_t* ptr, std::int64_t n) = 0;
+};
+
+struct inner_alloc {
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp:22]: struct -> class

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp [hunk 10]
```diff
@@ -12,6 +13,48 @@
...
+    byte_alloc_iface* get_byte_allocator() {
+        return byte_allocator;
+    }
+
+    const byte_alloc_iface* get_byte_allocator() const {
+        return byte_allocator;
+    }
+
+private:
+    byte_alloc_iface* byte_allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp]: byte_allocator -> byte_allocator_

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp [hunk 11]
```diff
@@ -184,6 +228,8 @@ class bit_vector {
                          const std::uint8_t* vector,
                          const std::int64_t vertex);
 
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp]: private ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp [hunk 12]
```diff
@@ -229,6 +276,7 @@ class graph {
 private:
     bool external_data;
     bool bit_representation;
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph.hpp]: allocator_

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph_matching_default_kernel.hpp [hunk 13]
```diff
@@ -24,54 +24,78 @@
...
 
+template <typename Alloc>
+struct alloc_connector : public byte_alloc_iface {
+    using byte_t = char;
+    using t_allocator_traits =
+        typename std::allocator_traits<Alloc>::template rebind_traits<byte_t>;
+    alloc_connector(Alloc alloc) : _alloc(alloc) {}
+    byte_t* allocate(std::int64_t count) override {
+        typename t_allocator_traits::pointer ptr = t_allocator_traits::allocate(_alloc, count);
+        if (ptr == nullptr) {
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph_matching_default_kernel.hpp:39]: What is the reason for that?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph_matching_default_kernel.hpp [hunk 14]
```diff
@@ -24,54 +24,78 @@
...
+            throw host_bad_alloc();
+        }
+        return ptr;
+    };
+
+    void deallocate(byte_t* ptr, std::int64_t count) override {
+        if (ptr != nullptr) {
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/graph_matching_default_kernel.hpp:46]: Do we need this 'if'?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp [hunk 15]
```diff
@@ -262,15 +268,15 @@ void matching_engine::push_into_stack(const std::int64_t vertex_id) {
...
+    state null_state(_allocator);
     std::int64_t candidates_count = 0;
     std::int64_t degree = pattern->get_vertex_degree(sorted_pattern_vertex[0]);
     for (std::int64_t i = 0; i < target->get_vertex_count(); i++) {
         if (degree <= target->get_vertex_degree(i) &&
             pattern->get_vertex_attribute(sorted_pattern_vertex[0]) ==
                 target->get_vertex_attribute(i)) {
-            void* place = _mm_malloc(sizeof(state), 64);
-            state* new_state = new (place) state(&null_state, i);
+            void* place = (void*)_allocator.allocate<state>(1);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp]: Why you need this explicit conversion?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp [hunk 16]
```diff
@@ -322,9 +328,9 @@ std::int64_t matching_engine::extract_candidates(state* current_state, bool chec
 bool matching_engine::check_vertex_candidate(const std::int64_t pattern_vertex,
                                              const std::int64_t target_vertex) {
     if (match_vertex(pattern_vertex, target_vertex)) {
-        state null_state;
-        void* place = _mm_malloc(sizeof(state), 64);
-        state* new_state = new (place) state(&null_state, target_vertex);
+        state null_state(_allocator);
+        void* place = (void*)_allocator.allocate<state>(1);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp]: Why you need this explicit conversion?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp [hunk 17]
```diff
@@ -333,8 +339,8 @@ bool matching_engine::check_vertex_candidate(const std::int64_t pattern_vertex,
 
 bool matching_engine::check_vertex_candidate(state* current_state, bool check_solution) {
     if (match_vertex(sorted_pattern_vertex[current_state->core_length], candidate)) {
-        void* place = _mm_malloc(sizeof(state), 64);
-        state* new_state = new (place) state(current_state, candidate);
+        void* place = (void*)_allocator.allocate<state>(1);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp]: Why you need this explicit conversion?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp [hunk 18]
```diff
@@ -428,8 +438,8 @@ solution engine_bundle::run() {
     }
 
     const std::uint64_t array_size = max_threads_count * 2;
-    matching_engine* engine_array =
-        static_cast<matching_engine*>(_mm_malloc(sizeof(matching_engine) * array_size, 64));
+    auto engine_array_ptr = _allocator.make_shared_memory<matching_engine>(array_size);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp]: const auto ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp [hunk 19]
```diff
@@ -428,8 +438,8 @@ solution engine_bundle::run() {
...
+    matching_engine* engine_array = engine_array_ptr.get();
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp:442]: * const ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp [hunk 20]
```diff
@@ -481,12 +491,13 @@ void engine_bundle::first_states_generator(bool use_exploration_stack) {
...
                                   [=](const int i) {
                                       bundle::ptr_t local_engine = matching_bundle.local();
-                                      state null_state;
+                                      state null_state(_allocator);
                                       if (degree <= target->get_vertex_degree(i) &&
                                           pattern->get_vertex_attribute(sorted_pattern_vertex[0]) ==
                                               target->get_vertex_attribute(i)) {
-                                          void* place = _mm_malloc(sizeof(state), 64);
-                                          state* new_state = new (place) state(&null_state, i);
+                                          void* place = (void*)_allocator.allocate<state>(1);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.cpp]: Why you need this explicit conversion?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.hpp [hunk 21]
```diff
@@ -48,6 +56,8 @@ class matching_engine {
     bool check_vertex_candidate(const std::int64_t pattern_vertex,
                                 const std::int64_t target_vertex);
 
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.hpp]: private ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.hpp [hunk 22]
```diff
@@ -91,10 +101,13 @@ class engine_bundle {
                   sconsistent_conditions const* pcconditions,
                   float* ppattern_vertex_probability,
                   const std::uint64_t _control_flags,
-                  kind isomorphism_kind);
+                  kind isomorphism_kind,
+                  inner_alloc _allocator);
     virtual ~engine_bundle();
     solution run();
 
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/matching.hpp]: private ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.cpp [hunk 23]
```diff
@@ -67,22 +69,23 @@ void solution::delete_data() {
     if (data != nullptr) {
         for (std::int64_t i = 0; i < max_solution_cout; i++) {
             if (data[i] != nullptr) {
-                _mm_free(data[i]);
+                _allocator.deallocate<std::int64_t>(data[i], 0);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.cpp]: Again 0

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.cpp [hunk 24]
```diff
@@ -138,7 +141,7 @@ graph_status solution::add(state* solution_state) {
         graph_status status = add(&solution_state->core);
         solution_state->~state();
         if (solution_state != nullptr) {
-            _mm_free(solution_state);
+            _allocator.deallocate<state>(solution_state, 0);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.cpp]: Again 0

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.cpp [hunk 25]
```diff
@@ -160,7 +163,7 @@ graph_status solution::add(solution& _solution) {
     }
 
     if (_solution.data != nullptr) {
-        _mm_free(_solution.data);
+        _allocator.deallocate<std::int64_t*>(_solution.data, 0);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.cpp]: Again 0

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.hpp [hunk 26]
```diff
@@ -34,6 +36,8 @@ class solution {
     graph_status add(solution&& _solution);
     oneapi::dal::homogen_table export_as_table();
 
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/solution.hpp]: private ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/sorter.hpp [hunk 27]
```diff
@@ -8,28 +8,34 @@ struct sconsistent_conditions {
     std::int64_t* array;
     std::int64_t divider;
     std::int64_t length;
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/sorter.hpp]: allocator_

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/sorter.hpp [hunk 28]
```diff
@@ -8,28 +8,34 @@ struct sconsistent_conditions {
...
 };
 
 class sorter {
 public:
-    sorter();
-    sorter(const graph* ptarget);
+    sorter(inner_alloc allocator);
+    sorter(const graph* ptarget, inner_alloc allocator);
     virtual ~sorter();
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/sorter.hpp]: private ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/stack.hpp [hunk 29]
```diff
@@ -43,6 +47,8 @@ class vertex_stack {
     std::uint64_t size() const;
     std::uint64_t max_size() const;
 
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/stack.hpp]: private ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/stack.hpp [hunk 30]
```diff
@@ -82,6 +92,8 @@ class dfs_stack {
 
     void delete_current_state();
 
+    inner_alloc _allocator;
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/stack.hpp]: private ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/stack.hpp [hunk 31]
```diff
@@ -55,10 +61,14 @@ class vertex_stack {
...
-    dfs_stack(const std::uint64_t levels, const std::uint64_t max_states_size);
-    dfs_stack(const std::uint64_t levels, const std::uint64_t* max_states_size_per_level);
+    dfs_stack(inner_alloc _allocator);
+    dfs_stack(const std::uint64_t levels, inner_alloc _allocator);
+    dfs_stack(const std::uint64_t levels,
+              const std::uint64_t max_states_size,
+              inner_alloc _allocator);
+    dfs_stack(const std::uint64_t levels,
+              const std::uint64_t* max_states_size_per_level,
+              inner_alloc _allocator);
```
Comment (averbukh) [inline cpp/oneapi/dal/algo/subgraph_isomorphism/detail/stack.hpp]: _allocator -> allocator ?

File: cpp/oneapi/dal/algo/subgraph_isomorphism/detail/debug.cpp (hunk 1 above)
Comment (orazve) [reply cpp/oneapi/dal/algo/subgraph_isomorphism/detail/debug.cpp:21]: yes

**Pull request merged**